File: src/main/java/quinzical/impl/questionparser/Question.java
package quinzical.impl.questionparser;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

public class Question implements Serializable {

    private static final long serialVersionUID = 1L;

    protected final String hint;
    protected final List<Solution> solutions = new ArrayList<>();
    protected final String prompt;

    public Question(String hint, String prompt) {
        this.hint = hint;
        this.prompt = prompt;
    }

    /**
     * Copy constructor
     *
     * @param q question to copy
     */
    public Question(Question q) {
        this.hint = q.hint;
        this.solutions.addAll(q.solutions);
        this.prompt = q.prompt;
    }

    public void addSolution(String[] solutions) {
        Solution s = new Solution(solutions);
        this.solutions.add(s);
    }
    

}

File: src/main/server/src/model/Questions/Question.ts
class QuestionImpl implements Question {
    private solution: string[];
    private question: string;
    private prompt: string;

    private constructor($solutions: string[], $question: string, $prompt: string) {
        this.solution = $solutions;
        this.question = $question;
        this.prompt = $prompt;
    }

    getSendableData(): SendableQuestionData {
        return {
            prompt: this.prompt,
            question: this.question,
        };
    }

    get Solution() {
        return this.solution;
    }

    get Question() {
        return this.question;
    }

    get Prompt() {
        return this.prompt;
    }

    static create({ prompt, question, solutions }: QuestionData): Question {
        return new QuestionImpl(solutions, question, prompt);
    }
}

interface Question {
    Solution: string[];
    Question: string;
    Prompt: string;

    getSendableData(): SendableQuestionData;
}

interface QuestionData extends SendableQuestionData {
    solutions: string[];
}

interface SendableQuestionData {
    question: string;
    prompt: string;
}

export { Question, QuestionImpl, SendableQuestionData };
